From: "Barak A. Pearlmutter" <barak+git@cs.nuim.ie>
Date: Sat, 29 Mar 2014 10:47:22 +0000
Subject: systrace.h

---
 systrace.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/systrace.h b/systrace.h
index f57cdb2..7434a3d 100644
--- a/systrace.h
+++ b/systrace.h
@@ -34,7 +34,7 @@
 #include <sys/queue.h>
 
 #ifndef __APPLE__
-#define _PATH_XSYSTRACE	"/usr/X11R6/bin/xsystrace"
+#define _PATH_XSYSTRACE	"/usr/lib/xsystrace/xsystrace"
 #else
 #define	_PATH_XSYSTRACE	"/Applications/Utilities/Cocoa-Systrace.app/Contents/MacOS/Cocoa-Systrace"
 #endif
