diff -Naur systrace-1.0/systrace.h systrace-1.0.new/systrace.h
--- systrace-1.5/systrace.h	2003-07-26 10:39:20.000000000 +0200
+++ systrace-1.5.new/systrace.h	2003-07-26 10:39:58.000000000 +0200
@@ -34,7 +34,7 @@
 #include <sys/queue.h>
 
 #ifndef __APPLE__
-#define _PATH_XSYSTRACE	"/usr/X11R6/bin/xsystrace"
+#define _PATH_XSYSTRACE	"/usr/lib/xsystrace/xsystrace"
 #else
 #define	_PATH_XSYSTRACE	"/Applications/Utilities/Cocoa-Systrace.app/Contents/MacOS/Cocoa-Systrace"
 #endif
